chacha20_poly1305/src/chacha20.rs

```diff
@@ -0,0 +1,460 @@
+    block_count: u32,
+    /// Interal byte offset index of the block_count.
+    seek_offset_bytes: usize,
+}
+
+impl ChaCha20 {
+    /// Make a new instance of ChaCha20 from an index in the keystream.
+    pub const fn new(key: Key, nonce: Nonce, seek: u32) -> Self {
+        let block_count = seek / 64;
+        let seek_offset_bytes = (seek % 64) as usize;
+        ChaCha20 { key, nonce, block_count, seek_offset_bytes }
+    }
+
+    /// Make a new instance of ChaCha20 from a block in the keystream.
+    pub const fn new_from_block(key: Key, nonce: Nonce, block: u32) -> Self {
+        ChaCha20 { key, nonce, block_count: block, seek_offset_bytes: 0 }
+    }
+
+    /// Apply the keystream to a buffer.
+    pub fn apply_keystream(&mut self, buffer: &mut [u8]) {
+        let num_full_blocks = buffer.len() / CHACHA_BLOCKSIZE;
+        for block in 0..num_full_blocks {
+            let keystream =
+                keystream_at_slice(self.key, self.nonce, self.block_count, self.seek_offset_bytes);
+            for (buffer_byte, keystream_byte) in buffer
+                [block * CHACHA_BLOCKSIZE..(block + 1) * CHACHA_BLOCKSIZE]
+                .iter_mut()
+                .zip(keystream.iter())
+            {
+                *buffer_byte ^= *keystream_byte
+            }
+            self.block_count += 1;
+        }
+        if buffer.len() % 64 > 0 {
+            let keystream =
+                keystream_at_slice(self.key, self.nonce, self.block_count, self.seek_offset_bytes);
+            for (buffer_byte, keystream_byte) in
+                buffer[num_full_blocks * CHACHA_BLOCKSIZE..].iter_mut().zip(keystream.iter())
+            {
+                *buffer_byte ^= *keystream_byte
+            }
+            self.block_count += 1;
+        }
+    }
+
+    /// Get the keystream block at a specified block.
+    pub fn get_keystream(&mut self, block: u32) -> [u8; 64] {
+        self.block(block);
+        keystream_at_slice(self.key, self.nonce, self.block_count, self.seek_offset_bytes)
+    }
+
+    /// Update the index of the keystream to the given byte.
+    pub fn seek(&mut self, seek: u32) {
+        self.block_count = seek / 64;
+        self.seek_offset_bytes = (seek % 64) as usize;
+    }
+
+    /// Update the index of the keystream to a block.
+    pub fn block(&mut self, block: u32) {
+        self.block_count = block;
+        self.seek_offset_bytes = 0;
+    }
+}
+
+fn keystream_at_slice(key: Key, nonce: Nonce, count: u32, seek: usize) -> [u8; 64] {
+    let mut keystream: [u8; 128] = [0; 128];
+    let (first_half, second_half) = keystream.split_at_mut(64);
+
+    let mut state = State::new(key, nonce, count);
+    state.chacha_block();
+    first_half.copy_from_slice(&state.keystream());
+
+    let mut state = State::new(key, nonce, count + 1);
+    state.chacha_block();
+    second_half.copy_from_slice(&state.keystream());
+
+    let seeked_keystream: [u8; 64] = keystream[seek..seek + 64].try_into().expect("infallable");
```

Comment:
If you can't come up with an informative message it's better to just use `unwrap`. But in this case the informative message is "slicing produces 64-byte slice".